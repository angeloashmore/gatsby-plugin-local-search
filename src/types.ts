import { PluginOptions as GatsbyPluginOptions, NodeInput, Node } from 'gatsby'
import { CreateOptions as FlexSearchCreateOptions } from 'flexsearch'

export type LunrPluginOptions = {
    k1?: number
    b?: number
  };

export interface PartialContext {
  nodeModel: {
    getNodeById: (input: { id: string; type?: string }) => Node
  }
}

export type IndexableDocument = Record<string, unknown>

export type Store = Record<string, unknown>

export enum NodeType {
  LocalSearch = 'LocalSearch',
}

export type Engine = 'flexsearch' | 'lunr'

export interface LocalSearchNodeInput extends NodeInput {
  name: string
  engine: Engine
  index: string
  store: Store
}

interface NormalizerInput {
  errors?: unknown
  data?: unknown
}

export interface PluginOptions extends GatsbyPluginOptions {
  name: string
  engine: Engine
  engineOptions?: FlexSearchCreateOptions
  lunrOptions?: any
  ref?: string
  index?: string[]
  store?: string[]
  query: string
  normalizer: (input: NormalizerInput) => IndexableDocument[]
}

