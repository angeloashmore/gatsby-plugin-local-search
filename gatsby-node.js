// Use compiled gatsby-node.js
module.exports = require('./dist/gatsby-node')
