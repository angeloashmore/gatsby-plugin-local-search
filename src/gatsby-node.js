import createNodeHelpers from 'gatsby-node-helpers'
import lunr from 'lunr'
import FlexSearch from 'flexsearch'
import * as R from 'ramda'
import lowerFirst from 'lodash.lowerfirst'

const TYPE_PREFIX = 'LocalSearch'
const TYPE_INDEX = 'Index'
const TYPE_STORE = 'Store'

const { generateTypeName, generateNodeId } = createNodeHelpers({
  typePrefix: TYPE_PREFIX,
})

// Returns an exported FlexSearch index using the provided documents, fields,
// and ref.
const createFlexSearchIndexExport = ({ documents, fields, ref }) => {
  const index = FlexSearch.create()

  documents.forEach(doc => index.add(doc[ref], JSON.stringify(R.pick(fields, doc))))

  return index.export()
}

// Returns an exported Lunr index using the provided documents, fields, and
// ref.
const createLunrIndexExport = ({ documents, fields, ref }) => {
  const index = lunr(function() {
    this.ref(ref)
    fields.forEach(x => this.field(x))
    documents.forEach(x => this.add(x))
  })

  return JSON.stringify(index)
}

// Returns an exported index using the provided engine, documents, fields, and
// ref. Throws if the provided engine is invalid.
const createIndexExport = ({ reporter, name, engine, ...args }) => {
  switch (engine) {
    case 'flexsearch':
      return createFlexSearchIndexExport(args)

    case 'lunr':
      return createLunrIndexExport(args)

    default:
      reporter.error(
        `The gatsby-plugin-local-search engine option for index "${name}" is invalid. Must be one of: flexsearch, lunr. The index will be null.`,
      )

      return null
  }
}

// Create index and store during createPages and save to cache. The cached
// values will be used in createResolvers.
export const createPages = async (
  { graphql, cache, reporter },
  { name, ref = 'id', store: storeFields, index: indexFields, query, normalizer, engine },
) => {
  const result = await graphql(query)
  if (result.errors) throw R.head(result.errors)

  const documents = await Promise.resolve(normalizer(result))
  if (R.isEmpty(documents))
    reporter.warn(
      `The gatsby-plugin-local-search query for index "${name}" returned no nodes. The index and store will be empty.`,
    )

  const fields = R.reject(R.equals(ref),
    indexFields || R.pipe(R.head, R.keys)(documents)
  )

  const index = createIndexExport({
    reporter,
    name,
    engine,
    documents,
    fields,
    ref,
  })

  // Default to all fields if storeFields is not provided
  const store = R.pipe(
    R.map(R.pick(storeFields || [...fields, ref])),
    R.indexBy(R.prop(ref)),
  )(documents)

  // Save to cache to use later in GraphQL resolver.
  await cache.set(generateNodeId(TYPE_INDEX, name), index)
  await cache.set(generateNodeId(TYPE_STORE, name), store)

  return
}

// Set the GraphQL type for LocalSearchIndex.
export const createSchemaCustomization = (
  { actions: { createTypes }, schema },
  { name },
) => {
  createTypes([
    schema.buildObjectType({
      name: generateTypeName(`${TYPE_INDEX} ${name}`),
      fields: {
        id: 'ID',
        engine: 'String',
        index: 'String',
        store: 'String',
      },
    }),
  ])
}

export const createResolvers = async (
  { actions: { createTypes }, createResolvers, cache, schema },
  { name, engine },
) => {
  createResolvers({
    Query: {
      [lowerFirst(generateTypeName(name))]: {
        type: generateTypeName(`${TYPE_INDEX} ${name}`),
        resolve: async (_parent, _args, context) => {
          const index = await cache.get(generateNodeId(TYPE_INDEX, name))
          const store = await cache.get(generateNodeId(TYPE_STORE, name))

          return {
            id: name,
            engine,
            index,
            store: JSON.stringify(store),
          }
        },
      },
    },
  })

  return
}
