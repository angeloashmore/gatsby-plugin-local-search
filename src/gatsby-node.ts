import path from 'path'
import fs from 'fs'
import lunr from 'lunr'
import FlexSearch from 'flexsearch'
import {
  GatsbyNode,
  CreatePagesArgs,
  CreateSchemaCustomizationArgs,
} from 'gatsby'
import { pick } from 'lodash'
import { pascalCase } from 'pascal-case'

import {
  IndexableDocument,
  NodeType,
  PluginOptions,
  Store,
  LocalSearchNodeInput,
} from './types'

const DEFAULT_REF = 'id'

const msg = (input: string) => `gatsby-plugin-local-search - ${input}`

const createFlexSearchIndexExport = (
  documents: IndexableDocument[],
  pluginOptions: PluginOptions,
): string => {
  const { ref = DEFAULT_REF, index: indexFields, engineOptions } = pluginOptions

  const index = FlexSearch.create<IndexableDocument>(engineOptions)

  documents.forEach((doc) => {
    const serializedDoc = JSON.stringify(
      indexFields ? pick(doc, indexFields) : doc,
    )
    // Using "as number" due to FlexSearch's types, but it could technically be
    // a string as well.
    index.add(doc[ref] as number, serializedDoc)
  })

  return index.export()
}

const createLunrIndexExport = (
  documents: IndexableDocument[],
  pluginOptions: PluginOptions,
): string => {
  const { ref = DEFAULT_REF, index: indexFields } = pluginOptions

  const fields =
    indexFields ?? (documents.length > 0 ? Object.keys(documents[0]) : [])

  const index = lunr(function () {
    this.ref(ref)
    fields.forEach((field) => this.field(field))
    documents.forEach((doc) => this.add(doc))
  })

  return JSON.stringify(index)
}

const createIndexExport = (
  documents: IndexableDocument[],
  pluginOptions: PluginOptions,
  gatsbyContext: CreatePagesArgs,
): string | void => {
  const { reporter } = gatsbyContext
  const { name, engine } = pluginOptions

  switch (engine) {
    case 'flexsearch':
      return createFlexSearchIndexExport(documents, pluginOptions)

    case 'lunr':
      return createLunrIndexExport(documents, pluginOptions)

    default:
      reporter.error(
        msg(
          `The engine option for index "${name}" is invalid. It must be one of: flexsearch, lunr. The index will not be created.`,
        ),
      )
  }
}

// Callback style is necessary since createPages cannot be async or return a
// Promise. At least, that's what GatsbyNode['createNodes'] says.
export const createPages = async (
  gatsbyContext: CreatePagesArgs,
  pluginOptions: PluginOptions,
): Promise<void> => {
  const {
    actions,
    graphql,
    reporter,
    createNodeId,
    createContentDigest,
  } = gatsbyContext
  const { createNode } = actions
  const {
    name,
    engine,
    ref = DEFAULT_REF,
    store: storeFields,
    query,
    normalizer,
  } = pluginOptions

  const result = await graphql(query)

  if (result.errors) {
    reporter.error(
      msg(
        'The provided GraphQL query contains errors. The index will not be created.',
      ),
      result.errors[0],
    )
    return
  }

  const documents = (await Promise.resolve(normalizer(result))) || []

  if (documents.length < 1)
    reporter.warn(
      msg(
        `The query for index "${name}" returned no nodes. The index and store will be empty.`,
      ),
    )

  const filteredDocuments = documents.filter(
    (doc) => doc[ref] !== undefined && doc[ref] !== null,
  )

  const index = createIndexExport(
    filteredDocuments,
    pluginOptions,
    gatsbyContext,
  )
  if (!index) return

  const store = filteredDocuments.reduce((acc, doc) => {
    acc[String(doc[ref])] = storeFields ? pick(doc, storeFields) : doc

    return acc
  }, {} as Store)

  const nodeType = pascalCase(`${NodeType.LocalSearch} ${name}`)
  const nodeId = createNodeId(name)

  const node: LocalSearchNodeInput = {
    id: nodeId,
    name,
    engine,
    index,
    store,
    internal: {
      type: nodeType,
      contentDigest: createContentDigest({ index, store }),
    },
  }

  createNode(node)
}

export const createSchemaCustomization: NonNullable<
  GatsbyNode['createSchemaCustomization']
> = async (
  gatsbyContext: CreateSchemaCustomizationArgs,
  pluginOptions: PluginOptions,
) => {
  const { actions, schema, reporter, pathPrefix } = gatsbyContext
  const { createTypes } = actions
  const { name } = pluginOptions

  const nodeType = pascalCase(`${NodeType.LocalSearch} ${name}`)

  createTypes([
    schema.buildObjectType({
      name: nodeType,
      fields: {
        name: {
          type: 'String!',
          description: 'The name of the index.',
        },
        engine: {
          type: 'String!',
          description: 'The search engine used to create the index.',
        },
        index: {
          type: 'String!',
          description: 'The search index created using the selected engine.',
        },
        store: {
          type: 'JSON!',
          description:
            'A JSON object used to map search results to their data.',
        },
        publicIndexURL: {
          type: 'String!',
          description:
            "Save the index to the site's static directory and return a public URL to it.",
          resolve: (node: LocalSearchNodeInput) => {
            const filename = `${node.internal.contentDigest}.index.txt`

            const publicPath = path.join(
              process.cwd(),
              'public',
              'static',
              filename,
            )

            if (!fs.existsSync(publicPath))
              fs.writeFile(publicPath, node.index, (err) => {
                if (err)
                  reporter.error(
                    msg(
                      `Could not save the index for "${name}" to ${publicPath}`,
                    ),
                  )
              })

            return `${pathPrefix}/static/${filename}`
          },
        },
        publicStoreURL: {
          type: 'String!',
          description:
            "Save the store to the site's static directory and return a public URL to it.",
          resolve: (node: LocalSearchNodeInput) => {
            const filename = `${node.internal.contentDigest}.store.json`

            const publicPath = path.join(
              process.cwd(),
              'public',
              'static',
              filename,
            )

            if (!fs.existsSync(publicPath))
              fs.writeFile(publicPath, JSON.stringify(node.store), (err) => {
                if (err)
                  reporter.error(
                    msg(
                      `Could not save the store for "${name}" to ${publicPath}`,
                    ),
                  )
              })

            return `${pathPrefix}/static/${filename}`
          },
        },
      },
      interfaces: ['Node'],
    }),
  ])
}
