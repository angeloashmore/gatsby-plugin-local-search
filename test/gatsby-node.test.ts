import { CreatePagesArgs } from 'gatsby'
import lunr from 'lunr'
import FlexSearch from 'flexsearch'

import { createPages } from '../src/gatsby-node'
import { PluginOptions, Engine } from '../src/gatsby-node'

const mockActions = {
  deletePage: jest.fn(),
  createPage: jest.fn(),
  deleteNode: jest.fn(),
  deleteNodes: jest.fn(),
  createNode: jest.fn(),
  touchNode: jest.fn(),
  createNodeField: jest.fn(),
  createParentChildLink: jest.fn(),
  setWebpackConfig: jest.fn(),
  replaceWebpackConfig: jest.fn(),
  setBabelOptions: jest.fn(),
  setBabelPlugin: jest.fn(),
  setBabelPreset: jest.fn(),
  createJob: jest.fn(),
  setJob: jest.fn(),
  endJob: jest.fn(),
  setPluginStatus: jest.fn(),
  createRedirect: jest.fn(),
  addThirdPartySchema: jest.fn(),
  createTypes: jest.fn(),
  createFieldExtension: jest.fn(),
}

const mockGatsbyContext: CreatePagesArgs & {
  traceId: 'initial-createPages'
} = {
  pathPrefix: 'pathPrefix',
  boundActionCreators: mockActions,
  actions: mockActions,
  loadNodeContent: jest.fn(),
  store: {
    dispatch: jest.fn(),
    subscribe: jest.fn(),
    getState: jest.fn(),
    replaceReducer: jest.fn(),
  },
  emitter: {
    addListener: jest.fn(),
    on: jest.fn(),
    once: jest.fn(),
    prependListener: jest.fn(),
    prependOnceListener: jest.fn(),
    removeListener: jest.fn(),
    off: jest.fn(),
    removeAllListeners: jest.fn(),
    setMaxListeners: jest.fn(),
    getMaxListeners: jest.fn(),
    listeners: jest.fn(),
    rawListeners: jest.fn(),
    emit: jest.fn(),
    eventNames: jest.fn(),
    listenerCount: jest.fn(),
  },
  getNodes: jest.fn(),
  getNode: jest.fn(),
  getNodesByType: jest.fn(),
  hasNodeChanged: jest.fn(),
  reporter: {
    stripIndent: jest.fn(),
    format: jest.fn(),
    setVerbose: jest.fn(),
    setNoColor: jest.fn(),
    panic: jest.fn(),
    panicOnBuild: jest.fn(),
    error: jest.fn(),
    uptime: jest.fn(),
    success: jest.fn(),
    verbose: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    log: jest.fn(),
    activityTimer: jest.fn(),
    createProgress: jest.fn(),
  },
  getNodeAndSavePathDependency: jest.fn(),
  cache: {
    getAndPassUp: jest.fn(),
    wrap: jest.fn(),
    set: jest.fn(),
    mset: jest.fn(),
    get: jest.fn(),
    mget: jest.fn(),
    del: jest.fn(),
    reset: jest.fn(),
  },
  createNodeId: jest.fn().mockReturnValue('createNodeId'),
  createContentDigest: jest.fn().mockReturnValue('createContentDigest'),
  tracing: {
    tracer: {},
    parentSpan: {},
    startSpan: jest.fn(),
  },
  traceId: 'initial-createPages',
  waitForCascadingActions: false,
  parentSpan: {},
  graphql: jest.fn(),
}

interface QueryResult {
  data: {
    allNode: {
      edges: TestNode[]
    }
  }
}

interface TestNode {
  id: string
  foo: string
}

const mockQueryResult: QueryResult = {
  data: {
    allNode: {
      edges: [
        { id: 'id1', foo: 'bar' },
        { id: 'id2', foo: 'needle' },
      ],
    },
  },
}

const pluginOptions: PluginOptions = {
  name: 'name',
  engine: Engine.FlexSearch,
  query: 'query',
  normalizer: ({ data }) =>
    (data as QueryResult['data']).allNode.edges.map(node => ({
      id: node.id,
      foo: node.foo,
    })),
  plugins: [],
}

beforeAll(() => {
  ;(mockGatsbyContext.graphql as jest.Mock).mockReturnValue(
    Promise.resolve(mockQueryResult),
  )
})

describe('createPages', () => {
  beforeEach(() => jest.clearAllMocks())

  test('creates FlexSearch index', async () => {
    await new Promise(res =>
      createPages!(mockGatsbyContext, pluginOptions, res),
    )

    const createNode = mockGatsbyContext.actions.createNode as jest.Mock
    const node = createNode.mock.calls[0][0]
    const exportedIndex = node.index
    const index = FlexSearch.create()
    index.import(exportedIndex)

    expect(mockGatsbyContext.actions.createNode).toMatchSnapshot()
    expect(index.search('needle')).toMatchSnapshot()
  })

  test('creates Lunr index', async () => {
    await new Promise(res =>
      createPages!(
        mockGatsbyContext,
        { ...pluginOptions, engine: Engine.Lunr },
        res,
      ),
    )

    const createNode = mockGatsbyContext.actions.createNode as jest.Mock
    const node = createNode.mock.calls[0][0]
    const exportedIndex = node.index
    const index = lunr.Index.load(JSON.parse(exportedIndex))

    expect(mockGatsbyContext.actions.createNode).toMatchSnapshot()
    expect(index.search('needle')).toMatchSnapshot()
  })

  test('creates types', async () => {
    await new Promise(res =>
      createPages!(mockGatsbyContext, pluginOptions, res),
    )

    expect(mockGatsbyContext.actions.createTypes).toMatchSnapshot()
  })
})
