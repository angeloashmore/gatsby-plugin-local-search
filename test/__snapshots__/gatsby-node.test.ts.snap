// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPages creates types 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
        type LocalSearchName extends Node @dontInfer {
          engine: String!
          index: String!
          store: JSON!
        }
      ",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createPages with flexsearch engine creates index 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "engine": "flexsearch",
        "id": "createNodeId",
        "index": "[[{\\"i\\":[\\"id1\\",\\"id2\\"],\\"id\\":[\\"id1\\",\\"id2\\"]},{\\"id1\\":[\\"id1\\"],\\"f\\":[\\"id1\\",\\"id2\\"],\\"fo\\":[\\"id1\\",\\"id2\\"],\\"foo\\":[\\"id1\\",\\"id2\\"],\\"id2\\":[\\"id2\\"]},{\\"b\\":[\\"id1\\"],\\"ba\\":[\\"id1\\"],\\"bar\\":[\\"id1\\"],\\"n\\":[\\"id2\\"],\\"ne\\":[\\"id2\\"],\\"nee\\":[\\"id2\\"],\\"need\\":[\\"id2\\"],\\"needl\\":[\\"id2\\"],\\"needle\\":[\\"id2\\"]},{},{},{},{},{},{}],{},[\\"@id1\\",\\"@id2\\"]]",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "LocalSearchName",
        },
        "store": Object {
          "id1": Object {
            "foo": "bar",
            "id": "id1",
          },
          "id2": Object {
            "foo": "needle",
            "id": "id2",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createPages with flexsearch engine creates index 2`] = `
Array [
  "id2",
]
`;

exports[`createPages with lunr engine creates index 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "engine": "lunr",
        "id": "createNodeId",
        "index": "{\\"version\\":\\"2.3.8\\",\\"fields\\":[\\"id\\",\\"foo\\"],\\"fieldVectors\\":[[\\"id/id1\\",[0,0.693]],[\\"foo/id1\\",[1,0.693]],[\\"id/id2\\",[2,0.693]],[\\"foo/id2\\",[3,0.693]]],\\"invertedIndex\\":[[\\"bar\\",{\\"_index\\":1,\\"id\\":{},\\"foo\\":{\\"id1\\":{}}}],[\\"id1\\",{\\"_index\\":0,\\"id\\":{\\"id1\\":{}},\\"foo\\":{}}],[\\"id2\\",{\\"_index\\":2,\\"id\\":{\\"id2\\":{}},\\"foo\\":{}}],[\\"needl\\",{\\"_index\\":3,\\"id\\":{},\\"foo\\":{\\"id2\\":{}}}]],\\"pipeline\\":[\\"stemmer\\"]}",
        "internal": Object {
          "contentDigest": "createContentDigest",
          "type": "LocalSearchName",
        },
        "store": Object {
          "id1": Object {
            "foo": "bar",
            "id": "id1",
          },
          "id2": Object {
            "foo": "needle",
            "id": "id2",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`createPages with lunr engine creates index 2`] = `
Array [
  Object {
    "matchData": Object {
      "metadata": Object {
        "needl": Object {
          "foo": Object {},
        },
      },
    },
    "ref": "id2",
    "score": 0.693,
  },
]
`;
